File: src/main/java/peterlavalle/NativeUnpack.java
package peterlavalle;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

public class NativeUnpack {
	public static final File folder;
	public static final String name = System.getProperty("os.name").replaceAll("^(\\w+)[^\\w].*$", "$1").toLowerCase();
	public static final String arch = System.getProperty("os.arch").toLowerCase();

	static {
		File unpackDir = null;
		try {
			unpackDir = File.createTempFile("unpacked", "shared-objects");

			if (!unpackDir.delete()) {
				throw new IOException();
			}

			if (!unpackDir.mkdirs()) {
				throw new IOException();
			}

			unpackDir.deleteOnExit();

			final String config = name + "/" + arch;

			String[] sharedObjects;
			if (config.equals("windows/amd64")) {
				sharedObjects = new String[]{
						"lwjgl64.dll",
						"OpenAL64.dll"
				};
			} else {
				throw new IOException("Unknown system configuration `" + config + "`");
			}

			final byte[] data = new byte[128];
			for (final String sharedObjectName : sharedObjects) {
				final File sharedObjectFile = new File(unpackDir, sharedObjectName);

				FileOutputStream outputStream = new FileOutputStream(sharedObjectFile);
				InputStream inputStream = ClassLoader.getSystemResourceAsStream(sharedObjectName);

				for (int read; -1 != (read = inputStream.read(data)); ) {
					outputStream.write(data, 0, read);
				}

				outputStream.close();
				inputStream.close();
				sharedObjectFile.deleteOnExit();
			}
		} catch (IOException e) {
			unpackDir = null;
			throw new RuntimeException(e);
		} finally {
			folder = unpackDir;
		}
	}

	private NativeUnpack() {
	}
}

File: neotextureedit.gradle/src/main/scala/peterlavalle/neo/NativeUnpack.scala
package peterlavalle.neo

import java.io.{File, FileOutputStream}

object NativeUnpack {
	val name: String = System.getProperty("os.name").replaceAll("^(\\w+)[^\\w].*$", "$1").toLowerCase
	val arch: String = System.getProperty("os.arch").toLowerCase

	val unpackDir: File = {
		val unpackDir: File =
			File.createTempFile("unpacked", "shared-objects")

		require(unpackDir.delete() && unpackDir.mkdirs())
		unpackDir.deleteOnExit()

		// TODO; more-better delete

		((name, arch) match {
			case ("windows", "amd64") =>
				List("lwjgl64.dll", "OpenAL64.dll")
		})
			.foreach {
				sharedObjectName =>

					// create the file
					val sharedObjectFile = new File(unpackDir, sharedObjectName).EnsureParent

					// fill the file with data
					(new FileOutputStream(sharedObjectFile) << ClassLoader.getSystemResourceAsStream(sharedObjectName))
						.close()

					// delete the file when we exit
					sharedObjectFile.deleteOnExit()
			}

		// add our path to java.library.path
		System.setProperty("java.library.path", unpackDir.getAbsolutePath + File.pathSeparator + System.getProperty("java.library.path"))

		// also; add us to the LWJGL path
		System.setProperty("org.lwjgl.librarypath", unpackDir.getAbsolutePath)

		// set sys_paths to null so that java.library.path will be reevalueted next time it is needed
		// ... this is an ugly necessity; we need to do this to kick the JVM
		// ... since this code is only ever called once; we're safe though - right?
		try {
			val sysPathsField = classOf[ClassLoader].getDeclaredField("sys_paths")
			sysPathsField.setAccessible(true)
			sysPathsField.set(null, null)
			unpackDir
		} catch {
			case e: NoSuchFieldException =>
				throw new RuntimeException("Something is missing", e)
			case e: IllegalAccessException =>
				throw new RuntimeException("I needed to modify the path for this to work", e)
		}
	}
}